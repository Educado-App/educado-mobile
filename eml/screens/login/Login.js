import React, { useEffect, useState } from 'react';
import { View, Keyboard } from 'react-native';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { useNavigation } from '@react-navigation/native';
import LoginForm from '../../components/login/LoginForm';
import LogoBackButton from '../../components/login/LogoBackButton';
import { SafeAreaView } from 'react-native-safe-area-context';
import { TouchableWithoutFeedback } from 'react-native';
import Text from '../../components/general/Text';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';
import LoadingScreen from '../../components/loading/Loading';
import * as StorageService from '../../services/StorageService';

const LOGIN_TOKEN = '@loginToken';

/**
 * Login screen component containing a login form and possibilities of resetting password or registering a new user.
 * @param {Object} props not used in this component as of now
 */
export default function Login() {

	const [loading, setLoading] = useState(true);
	const navigation = useNavigation();

	/**
   * TODO: Refactor error to use new error handling system
   * Function for checking if a login token is stored in async local storage (i.e. if the user is already logged in)
   * If a token is found, the user is redirected to the home screen.
   * 
   */
	const checkLoginToken = async () => {
		try {
			const fetchedToken = await AsyncStorage.getItem(LOGIN_TOKEN);
			if (fetchedToken !== null) {
				await AsyncStorage.setItem('loggedIn', 'true');
				StorageService.updateStoredCourses();
				navigation.navigate('HomeStack');
			}
			setLoading(false);
		} catch (error) {
			console.log('Failed to fetch the login token from storage');
			setLoading(false);
		}
	};

	useEffect(() => {
		checkLoginToken();
	}, []);

	return (
		<SafeAreaView className="justify-start bg-secondary flex-1">
			{loading ? (<LoadingScreen />) :
				(<KeyboardAwareScrollView
					className="flex-1"
					resetScrollToCoords={{ x: 0, y: 0 }}
					scrollEnabled={true}>
					<TouchableWithoutFeedback onPress={Keyboard.dismiss}>
						<View>
							<View className="mt-10">
								<LogoBackButton navigationPlace="WelcomeStack" />
							</View>
							<View className="mx-6">
								{/* Login form */}
								<View className="my-8">
									<LoginForm />
								</View>
								{/* Register button */}
								<View className="flex-row justify-center">
									<Text className="text-base text-projectGray mr-1">
										{/* Dont have an account yet? */}
                    Ainda n√£o tem conta?
									</Text>
									<Text
										testId="registerNav"
										className={'text-base text-projectBlack underline'}
										onPress={() => navigation.navigate('Register')}
									>
										{/* Sign up now */}
                    Cadastre-se agora
									</Text>
								</View>
							</View>
						</View>
					</TouchableWithoutFeedback>
				</KeyboardAwareScrollView>)}
		</SafeAreaView >
	);
}
