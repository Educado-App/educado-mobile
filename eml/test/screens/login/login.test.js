import React from 'react';
import renderer from 'react-test-renderer';
import Login from '../../../screens/login/Login';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';

let navigated = false;

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    navigate: jest.fn(() => { navigated = true; }),
  }),
}));

describe('Login screen', () => {

  let loginScreen;

  beforeEach(async () => {
    navigated = false;
    await AsyncStorage.clear();
    renderer.act(() => {
      loginScreen = renderer.create(<Login />);
    });
  });

  it('Login screen renders', () => {
    expect(loginScreen.toJSON()).toMatchSnapshot();
  });

  it('Pressing register new user navigates to the register page', async () => {
    const registerNav = loginScreen.root.findByProps({ testId: 'registerNav' });
    await renderer.act(() => {
      registerNav.props.onPress();
    });
    expect(navigated).toBe(true);
  });

  it('Check login when no valid token is stored', async () => {
    await renderer.act(() => {
      renderer.create(<Login />);
    });
    expect(navigated).toBe(false);
  });

  it('Check login when valid token stored', async () => {
    AsyncStorage.setItem('@loginToken', 'testToken');
    await renderer.act(() => {
      renderer.create(<Login />);
    });
    expect(navigated).toBe(true);
  });

  it('Check screen is scrollable with keyboard active', async () => {
    const scrollView = loginScreen.root.findByType(KeyboardAwareScrollView);
    expect(scrollView.props.scrollEnabled).toBeTruthy();
  });
});

/* TODO: Fix tests with AsyncStorage */ /*
test('Check login when valid token stored', async () => {
  const mockToken = "testToken";
  AsyncStorage.setItem("@loginToken", mockToken).then(async () => {
    await AsyncStorage.getItem("@loginToken");
    renderer.create(<Login />);
    expect(useNavigation().navigate).toHaveBeenCalledTimes(1);
  });
  
})*/