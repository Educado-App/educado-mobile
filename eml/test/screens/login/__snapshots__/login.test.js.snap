// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login screen Login screen renders 1`] = `
<RNCSafeAreaView
  edges={
    [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    [
      {
        "justifyContent": "flex-start",
      },
      {
        "backgroundColor": "#F1F9FB",
      },
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      style={
        [
          {
            "marginTop": 40,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "flexDirection": "row",
            },
            {
              "justifyContent": "center",
            },
            {
              "alignItems": "center",
            },
            {
              "width": "100%",
            },
            {
              "marginTop": 16,
            },
          ]
        }
      >
        <View
          style={
            [
              {
                "position": "absolute",
              },
              {
                "left": 0,
              },
              {
                "zIndex": 50,
              },
            ]
          }
        >
          <View
            style={
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderRadius": 20,
                },
                undefined,
                false,
              ]
            }
            testID="RNE_BUTTON_WRAPPER"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "busy": false,
                  "disabled": false,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "invisible",
                    "borderColor": "#2089dc",
                    "borderRadius": 20,
                    "borderWidth": 0,
                    "flexDirection": "row",
                    "height": 50,
                    "justifyContent": "center",
                    "padding": 4,
                    "paddingHorizontal": 6,
                    "width": 50,
                  }
                }
              >
                <View
                  style={
                    {
                      "overflow": "hidden",
                    }
                  }
                  testID="RNE__ICON__CONTAINER"
                >
                  <View
                    accessibilityRole="button"
                    android_ripple={
                      {
                        "borderless": false,
                        "color": "rgba(0, 0, 0, 0.3)",
                        "radius": -5,
                      }
                    }
                    disabled={false}
                    testID="RNE__ICON__CONTAINER_ACTION"
                  >
                    <View
                      style={
                        {
                          "alignItems": "center",
                          "backgroundColor": "transparent",
                          "justifyContent": "center",
                        }
                      }
                      testID="RNE__ICON"
                    >
                      <Text
                        allowFontScaling={false}
                        selectable={false}
                        style={
                          [
                            {
                              "color": "#383838",
                              "fontSize": 25,
                            },
                            {
                              "backgroundColor": "transparent",
                            },
                            {
                              "fontFamily": "material-community",
                              "fontStyle": "normal",
                              "fontWeight": "normal",
                            },
                            {},
                          ]
                        }
                        testID="RNE__ICON__Component"
                      >
                        Û∞ÖÅ
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            [
              {
                "alignItems": "center",
              },
              {
                "justifyContent": "center",
              },
              {
                "width": "100%",
              },
            ]
          }
        >
          <Image
            resizeMode="contain"
            source={1}
            style={
              [
                {
                  "height": 48,
                },
              ]
            }
          />
        </View>
      </View>
    </View>
    <View
      style={
        [
          {
            "marginLeft": 24,
            "marginRight": 24,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "marginBottom": 32,
              "marginTop": 32,
            },
          ]
        }
      >
        <View>
          <View
            style={
              [
                {
                  "marginBottom": 24,
                },
              ]
            }
          >
            <View>
              <View
                style={
                  [
                    {
                      "display": "flex",
                    },
                    {
                      "flexDirection": "row",
                    },
                  ]
                }
              >
                <Text
                  style={
                    [
                      {
                        "fontFamily": "Montserrat-Regular",
                      },
                      {
                        "color": "#383838",
                      },
                      {
                        "fontSize": 16,
                      },
                      [
                        {
                          "marginLeft": 8,
                        },
                        {
                          "fontSize": 12,
                          "lineHeight": 16,
                        },
                      ],
                    ]
                  }
                >
                  E-mail
                </Text>
                <Text
                  style={
                    [
                      {
                        "fontFamily": "Montserrat-Regular",
                      },
                      {
                        "color": "#383838",
                      },
                      {
                        "fontSize": 16,
                      },
                      [
                        {
                          "marginLeft": 4,
                        },
                        {
                          "fontSize": 12,
                          "lineHeight": 16,
                        },
                        {
                          "color": "#FF4949",
                        },
                      ],
                    ]
                  }
                >
                  *
                </Text>
              </View>
              <View>
                <TextInput
                  autoComplete="off"
                  keyboardType="email-address"
                  onChangeText={[Function]}
                  passwordGuidelines={false}
                  placeholder="Insira sua e-mail"
                  secureTextEntry={false}
                  style={
                    [
                      {
                        "paddingBottom": 4,
                        "paddingTop": 4,
                      },
                      {
                        "paddingLeft": 10,
                      },
                      {
                        "backgroundColor": "#FFFFFF",
                      },
                      {
                        "borderBottomLeftRadius": 8,
                        "borderBottomRightRadius": 8,
                        "borderTopLeftRadius": 8,
                        "borderTopRightRadius": 8,
                      },
                    ]
                  }
                />
              </View>
            </View>
            <View
              style={
                [
                  {
                    "flexDirection": "row",
                  },
                  {
                    "alignItems": "center",
                  },
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "fontFamily": "Montserrat-Regular",
                    },
                    {
                      "color": "#383838",
                    },
                    {
                      "fontSize": 16,
                    },
                    [
                      {
                        "fontSize": 12,
                        "lineHeight": 16,
                      },
                      {
                        "color": "#FF4949",
                      },
                      {
                        "marginLeft": 8,
                        "marginRight": 8,
                      },
                    ],
                  ]
                }
              />
            </View>
          </View>
          <View
            style={
              [
                {
                  "position": "relative",
                },
                {
                  "marginBottom": 24,
                },
              ]
            }
          >
            <View>
              <View
                style={
                  [
                    {
                      "display": "flex",
                    },
                    {
                      "flexDirection": "row",
                    },
                  ]
                }
              >
                <Text
                  style={
                    [
                      {
                        "fontFamily": "Montserrat-Regular",
                      },
                      {
                        "color": "#383838",
                      },
                      {
                        "fontSize": 16,
                      },
                      [
                        {
                          "marginLeft": 8,
                        },
                        {
                          "fontSize": 12,
                          "lineHeight": 16,
                        },
                      ],
                    ]
                  }
                >
                  Senha
                </Text>
                <Text
                  style={
                    [
                      {
                        "fontFamily": "Montserrat-Regular",
                      },
                      {
                        "color": "#383838",
                      },
                      {
                        "fontSize": 16,
                      },
                      [
                        {
                          "marginLeft": 4,
                        },
                        {
                          "fontSize": 12,
                          "lineHeight": 16,
                        },
                        {
                          "color": "#FF4949",
                        },
                      ],
                    ]
                  }
                >
                  *
                </Text>
              </View>
              <View>
                <TextInput
                  autoComplete="off"
                  keyboardType="default"
                  onChangeText={[Function]}
                  passwordGuidelines={false}
                  placeholder="Insira sua senha"
                  secureTextEntry={true}
                  style={
                    [
                      {
                        "paddingBottom": 4,
                        "paddingTop": 4,
                      },
                      {
                        "paddingLeft": 10,
                      },
                      {
                        "backgroundColor": "#FFFFFF",
                      },
                      {
                        "borderBottomLeftRadius": 8,
                        "borderBottomRightRadius": 8,
                        "borderTopLeftRadius": 8,
                        "borderTopRightRadius": 8,
                      },
                    ]
                  }
                  value=""
                />
              </View>
            </View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  {
                    "position": "absolute",
                  },
                  {
                    "right": 0,
                  },
                  {
                    "top": 8,
                  },
                  {
                    "paddingBottom": 12,
                    "paddingLeft": 12,
                    "paddingRight": 12,
                    "paddingTop": 12,
                  },
                ]
              }
            >
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "gray",
                      "fontSize": 24,
                    },
                    undefined,
                    {
                      "fontFamily": "material-community",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                Û∞àà
              </Text>
            </View>
            <View
              style={
                [
                  {
                    "flexDirection": "row",
                  },
                  {
                    "alignItems": "center",
                  },
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "fontFamily": "Montserrat-Regular",
                    },
                    {
                      "color": "#383838",
                    },
                    {
                      "fontSize": 16,
                    },
                    [
                      {
                        "fontSize": 12,
                        "lineHeight": 16,
                      },
                      {
                        "color": "#FF4949",
                      },
                      {
                        "marginLeft": 8,
                        "marginRight": 8,
                      },
                    ],
                  ]
                }
              />
            </View>
          </View>
          <View>
            <Text
              onPress={[Function]}
              style={
                [
                  {
                    "fontFamily": "Montserrat-Regular",
                  },
                  {
                    "color": "#383838",
                  },
                  {
                    "fontSize": 16,
                  },
                  [
                    {
                      "textAlign": "right",
                    },
                    {
                      "textDecorationLine": "underline",
                    },
                    {
                      "fontSize": 16,
                      "lineHeight": 24,
                    },
                    {
                      "color": "#383838",
                    },
                  ],
                ]
              }
            >
              Esqueceu a senha?
            </Text>
          </View>
          <View>
            <View
              accessibilityState={
                {
                  "disabled": true,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "backgroundColor": "#5ECCE9",
                  "borderBottomLeftRadius": 8,
                  "borderBottomRightRadius": 8,
                  "borderTopLeftRadius": 8,
                  "borderTopRightRadius": 8,
                  "marginTop": 16,
                  "opacity": 0.5,
                  "paddingBottom": 16,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "paddingTop": 16,
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "fontFamily": "Montserrat-Regular",
                    },
                    {
                      "color": "#383838",
                    },
                    {
                      "fontSize": 16,
                    },
                    [
                      {
                        "textAlign": "center",
                      },
                      {
                        "fontFamily": "Montserrat-Bold",
                      },
                      {
                        "fontSize": 16,
                      },
                      {
                        "color": "#FFFFFF",
                      },
                    ],
                  ]
                }
              >
                Entrar
              </Text>
            </View>
          </View>
          <View
            style={
              [
                {
                  "paddingTop": 40,
                },
              ]
            }
          >
            <Modal
              animationType="slide"
              hardwareAccelerated={false}
              style={
                [
                  {
                    "borderBottomWidth": 8,
                    "borderLeftWidth": 8,
                    "borderRightWidth": 8,
                    "borderTopWidth": 8,
                  },
                ]
              }
              visible={false}
            />
          </View>
        </View>
      </View>
      <View
        style={
          [
            {
              "flexDirection": "row",
            },
            {
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "Montserrat-Regular",
              },
              {
                "color": "#383838",
              },
              {
                "fontSize": 16,
              },
              [
                {
                  "fontSize": 16,
                  "lineHeight": 24,
                },
                {
                  "color": "#A1ACB2",
                },
                {
                  "marginRight": 4,
                },
              ],
            ]
          }
        >
          Ainda n√£o tem conta?
        </Text>
        <Text
          onPress={[Function]}
          style={
            [
              {
                "fontFamily": "Montserrat-Regular",
              },
              {
                "color": "#383838",
              },
              {
                "fontSize": 16,
              },
              [
                {
                  "fontSize": 16,
                  "lineHeight": 24,
                },
                {
                  "color": "#383838",
                },
                {
                  "textDecorationLine": "underline",
                },
              ],
            ]
          }
          testId="registerNav"
        >
          Cadastre-se agora
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
